Mangling: mangle box field types as they are.

eeckstein opened the pull request:
Instead of trying to mangle a tuple type.
This change avoids mangling tuple element names in case a field type is a tuple with named elements.
Fixes a "can't demangle" crash in the closure specializer.
    
This mangling is only used for specializations, so no ABI effect.

rdar://problem/46380088

eeckstein commented:
@swift-ci test

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -954,15 +954,15 @@ void ASTMangler::appendType(Type type) {
       }
       appendTypeList(TupleType::get(fieldsList, tybase->getASTContext())
-                       ->getCanonicalType());
+                       ->getCanonicalType(), /*ignoreElementNames=*/true);
```
There are no field names here anyway by construction

eeckstein replied:
If there is a field which itself is a tuple type (with field names), then getCanonicalType will flatten the tuple and it will have field names

eeckstein replied:
I don't see why creating a new tuple type would be better than passing a flag

jrose-apple replied:
Because then we don't have to support two modes in the mangler. I'm with Slava on this.

eeckstein replied:
That doesn't make sense to me.
Why should I create a new type (which involves memory allocation etc.) instead of just passing a boolean flag?
Maybe it's better to duplicate appendTypeList for both purposes (I though it is not) instead of passing a flag, but creating a temporary type for that is too much overhead.

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -954,15 +954,15 @@ void ASTMangler::appendType(Type type) {
         }
         appendTypeList(TupleType::get(fieldsList, tybase->getASTContext())
-                         ->getCanonicalType());
+                         ->getCanonicalType(), /*ignoreElementNames=*/true);
```
Or here

slavapestov commented on include/swift/AST/ASTMangler.h:
```diff
@@ -241,7 +241,7 @@ class ASTMangler : public Mangler {
   void appendFunctionResultType(Type resultType);
 
-  void appendTypeList(Type listTy);
+  void appendTypeList(Type listTy, bool ignoreElementNames = false);
```
I don't see anywhere this is set to 'true'.

slavapestov reviewed (changes requested):
I don't see anywhere you pass 'true' for the new parameter. And wouldn't it be better to build a new TupleType without labels instead of plumbing through a new boolean flag?

eeckstein commented:
What we can do is to just mangle the types - as they are - in a type_list, and not attempt to flatten top-level tuples. This would fix the element-name problem and the code would be clean.

jrose-apple commented:
That doesn't fix the element-name problem either, if you're worried about nested element names.

eeckstein committed:
Mangling: mangle box field types as they are.

Instead of trying to mangle a tuple type.
This change avoids mangling tuple element names in case a field type is a tuple with named elements.
Fixes a "can't demangle" crash in the closure specializer.

This mangling is only used for specializations, so no ABI effect.

rdar://problem/46380088

eeckstein commented:
It fixes it (as the test shows). The problem was that we mangled element types + their names not within a tuple context ('t').

eeckstein commented:
I pushed a new version

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
yes, it can

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test macOS

slavapestov reviewed (approved):
This feels like a better fix. Thanks for digging deeper here.

Do you need to update ABI.rst? Also the PR description/title is now out of date.

eeckstein commented:
No need to update ABI.rst. I updated the PR description, thanks.

eeckstein commented:
@swift-ci smoke test macOS

eeckstein commented:
@swift-ci smoke test macOS

eeckstein commented:
@swift-ci smoke test macos

eeckstein commented:
@swift-ci smoke test macos

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test macos
